rust/serving/src/source.rs

```diff
@@ -0,0 +1,269 @@
+pub(crate) struct MessageWrapper {
+    // TODO: this might be more that saving to ISB.
+    pub(crate) confirm_save: oneshot::Sender<()>,
+    pub(crate) message: Message,
+}
+
+/// Serving payload passed on to Numaflow.
+#[derive(Debug)]
+pub struct Message {
+    pub value: Bytes,
+    pub id: String,
+    pub headers: HashMap<String, String>,
+}
+
+enum ActorMessage {
+    Read {
+        batch_size: usize,
+        timeout_at: Instant,
+        reply_to: oneshot::Sender<Vec<Message>>,
+    },
+    Ack {
+        offsets: Vec<String>,
+        reply_to: oneshot::Sender<()>,
+    },
+}
+
+struct ServingSourceActor {
+    messages: mpsc::Receiver<MessageWrapper>,
+    handler_rx: mpsc::Receiver<ActorMessage>,
+    tracker: HashMap<String, oneshot::Sender<()>>,
+}
+
+impl ServingSourceActor {
+    async fn start(
+        settings: Arc<Settings>,
+        handler_rx: mpsc::Receiver<ActorMessage>,
+    ) -> Result<()> {
+        let (messages_tx, messages_rx) = mpsc::channel(10000);
+        // Create a redis store to store the callbacks and the custom responses
+        let redis_store = RedisConnection::new(settings.redis.clone()).await?;
+        // Create the message graph from the pipeline spec and the redis store
+        let msg_graph = MessageGraph::from_pipeline(&settings.pipeline_spec).map_err(|e| {
+            Error::InitError(format!(
+                "Creating message graph from pipeline spec: {:?}",
+                e
+            ))
+        })?;
+        let callback_state = CallbackState::new(msg_graph, redis_store).await?;
+
+        tokio::spawn(async move {
+            let mut serving_actor = ServingSourceActor {
+                messages: messages_rx,
+                handler_rx,
+                tracker: HashMap::new(),
+            };
+            serving_actor.run().await;
+        });
+        let app = crate::AppState {
+            message: messages_tx,
+            settings,
+            callback_state,
+        };
+        tokio::spawn(async move {
+            crate::serve(app).await.unwrap();
+        });
+        Ok(())
+    }
+
+    async fn run(&mut self) {
+        while let Some(msg) = self.handler_rx.recv().await {
+            self.handle_message(msg).await;
+        }
+    }
+
+    async fn handle_message(&mut self, actor_msg: ActorMessage) {
+        match actor_msg {
+            ActorMessage::Read {
+                batch_size,
+                timeout_at,
+                reply_to,
+            } => {
+                let messages = self.read(batch_size, timeout_at).await;
+                let _ = reply_to.send(messages);
+            }
+            ActorMessage::Ack { offsets, reply_to } => {
+                self.ack(offsets).await;
+                let _ = reply_to.send(());
+            }
+        }
+    }
+
+    async fn read(&mut self, count: usize, timeout_at: Instant) -> Vec<Message> {
+        let mut messages = vec![];
+        loop {
```

Comment:
return `Result<Vec<Message>>` to fix the FIXME :).  Source reader trait returns a result.